src/datasources/db/postgres-database.migrator.ts

```diff
@@ -0,0 +1,233 @@
+ *
+ * This is heavily inspired by `postgres-shift`
+ * @see https://github.com/porsager/postgres-shift/blob/master/index.js
+ */
+@Injectable()
+export class PostgresDatabaseMigrator {
+  private static readonly MIGRATIONS_FOLDER = join(process.cwd(), 'migrations');
+  private static readonly SQL_MIGRATION_FILE = 'index.sql';
+  private static readonly JS_MIGRATION_FILE = 'index.js';
+  private static readonly MIGRATIONS_TABLE = 'migrations';
+
+  constructor(@Inject('DB_INSTANCE') private readonly sql: Sql) {}
+
+  /**
+   * Runs/records migrations not present in the {@link PostgresMigrator.MIGRATIONS_TABLE} table.
+   *
+   * Note: all migrations are run in a single transaction for optimal performance.
+   */
+  async migrate(
+    path = PostgresDatabaseMigrator.MIGRATIONS_FOLDER,
+  ): Promise<void> {
+    const migrations = this.getMigrations(path);
+
+    await this.assertMigrationsTable();
+
+    const last = await this.getLastRunMigration();
+    const remaining = migrations.slice(last?.id ?? 0);
+
+    await this.sql.begin(async (transaction: TransactionSql) => {
+      for (const current of remaining) {
+        await this.run({ transaction, migration: current });
+        await this.setLastRunMigration({ transaction, migration: current });
+      }
+    });
+  }
+
+  /**
+   * @private migrates up to/allows for querying before/after migration to test it.
+   *
+   * Note: each migration is ran in separate transaction to allow queries in between.
+   *
+   * @param args.migration - migration to test
+   * @param args.folder - folder to search for migrations
+   * @param args.before - function to run before each migration
+   * @param args.after - function to run after each migration
+   *
+   * @example
+   * ```typescript
+   * const result = await migrator.test({
+   *   migration: '00001_initial',
+   *   before: (sql) => sql`SELECT * FROM <table_name>`,
+   *   after: (sql) => sql`SELECT * FROM <table_name>`,
+   * });
+   *
+   * expect(result.before).toBeUndefined();
+   * expect(result.after).toStrictEqual(expected);
+   * ```
+   */
+  async test(args: {
+    migration: string;
+    before?: (sql: Sql) => Promise<unknown>;
+    after: (sql: Sql) => Promise<unknown>;
+    folder?: string;
+  }): Promise<{
+    before: unknown;
+    after: unknown;
+  }> {
+    const migrations = this.getMigrations(
+      args.folder ?? PostgresDatabaseMigrator.MIGRATIONS_FOLDER,
+    );
+
+    let before: unknown;
+
+    shift: for await (const migration of migrations) {
+      const isMigrationBeingTested = migration.path.includes(args.migration);
+
+      if (isMigrationBeingTested && args.before) {
+        before = await args.before(this.sql).catch(() => undefined);
+      }
+
+      await this.sql.begin((transaction) => {
+        return this.run({ transaction, migration });
+      });
+
+      if (isMigrationBeingTested) {
+        // Exit loop after testing migration
+        break shift;
+      }
+    }
```

Comment:
Nitpick: the `shift` label is not needed, right? There's just one loop, so the `break;` statement would do the job.

Anyway, if labelling makes it more readable for you I'm fine with it.

Reply:
In f994465bd3345e2870f0ce090c90c6b516bbcead I changed it to only test up to the specified migration.